src/paths/steiner.cpp

```diff
@@ -0,0 +1,480 @@
+			
+
+				std::set<igraph_integer_t> F = fetchSetsBasedonIndex(holder - i);
+
+				std::set<igraph_integer_t> E;
+
+				std::set_difference(D.begin(),D.end(),F.begin(),F.end(),std::inserter(E, E.end()));
+				
+
+				igraph_real_t temp_value = MATRIX(*dp_cache,*E.begin(),k) + MATRIX(*dp_cache,holder - i,k) ;
+				
+				if (temp_value < min_value) {
+					min_value = temp_value;
+					min_E_value = *E.begin();
+					min_F = F;
+				}
+			}
+
+			igraph_vector_int_t vectorlist_1;
+			IGRAPH_CHECK(igraph_vector_int_init(&vectorlist_1,1));
+			IGRAPH_FINALLY(igraph_vector_int_destroy,&vectorlist_1);
+
+			igraph_vector_int_t edgelist_1;
+			IGRAPH_CHECK(igraph_vector_int_init(&edgelist_1,1));
+			IGRAPH_FINALLY(igraph_vector_int_destroy,&edgelist_1);
+
+			igraph_get_shortest_path_dijkstra(graph,&vectorlist_1,&edgelist_1,k,min_E_value,weights,IGRAPH_ALL);
+			
+			
+			igraph_vector_int_append(vectorlist_all,&vectorlist_1);
+			igraph_vector_int_append(edgelist_all,&edgelist_1);
+
+		
+			igraph_vector_int_destroy(&vectorlist_1);
+			igraph_vector_int_destroy(&edgelist_1);
+			IGRAPH_FINALLY_CLEAN(2);
+
+		}
+		else {
+			
+			igraph_integer_t E1,F1;
+
+			E1 = *D.begin();
+			F1 = *next(D.begin(),1);;	
+			
+
+
+			igraph_vector_int_t vectorlist_1;
+			IGRAPH_CHECK(igraph_vector_int_init(&vectorlist_1,1));
+			IGRAPH_FINALLY(igraph_vector_int_destroy,&vectorlist_1);
+
+			igraph_vector_int_t edgelist_1;
+			IGRAPH_CHECK(igraph_vector_int_init(&edgelist_1,1));
+			IGRAPH_FINALLY(igraph_vector_int_destroy,&edgelist_1);
+
+			igraph_get_shortest_path_dijkstra(graph,&vectorlist_1,&edgelist_1,k,E1,weights,IGRAPH_ALL);
+
+			igraph_vector_int_t vectorlist_2;
+			IGRAPH_CHECK(igraph_vector_int_init(&vectorlist_2,1));
+			IGRAPH_FINALLY(igraph_vector_int_destroy,&vectorlist_2);
+
+			igraph_vector_int_t edgelist_2;
+			IGRAPH_CHECK(igraph_vector_int_init(&edgelist_2,1));
+			IGRAPH_FINALLY(igraph_vector_int_destroy,&edgelist_2);
+
+			igraph_get_shortest_path_dijkstra(graph,&vectorlist_2,&edgelist_2,k,F1,weights,IGRAPH_ALL);
+
+			igraph_vector_int_append(vectorlist_all,&vectorlist_1);
+			igraph_vector_int_append(vectorlist_all,&vectorlist_2);
+
+			igraph_vector_int_append(edgelist_all,&edgelist_1);
+			igraph_vector_int_append(edgelist_all,&edgelist_2);
+
+			igraph_vector_int_destroy(&vectorlist_2);
+			igraph_vector_int_destroy(&vectorlist_1);
+			
+			igraph_vector_int_destroy(&edgelist_1);
+			igraph_vector_int_destroy(&edgelist_2);
+
+			IGRAPH_FINALLY_CLEAN(4);
+
+			std::set<igraph_integer_t> min_F;
+			min_F.insert(F1); 
+
+
+
+		}
+		
+		m = k;
+		D = min_F;			
+		
+		igraph_vector_int_destroy(&vectorlist);
+		igraph_vector_int_destroy(&edgelist);
+		IGRAPH_FINALLY_CLEAN(2);
+	}
+	
+
+
+	return IGRAPH_SUCCESS;
+
+}
+
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph,const igraph_vector_int_t* steiner_terminals,
+igraph_neimode_t mode, const igraph_vector_t *weights,igraph_real_t *res,igraph_vector_int_t *res_tree)
```

Comment:
Currently you only implemented the undirected case. We will need to decide if this function will even be generalized to the directed case. If not, it will be better to remove the `mode` argument entirely and document that the function ignores edge directions.

Correct me if I'm wrong, but it seems to me that the directed Steiner tree problem is conceptually different from the undirected one. The variants I found start with a specific root node, which we don't need for the undirected case. Does the Dreyfus-Wagner algorithm generalize to the directed case at all? If not, we do not need to consider the future possibility of directed genertalizations and change the function signature accordingly.